src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -182,31 +183,59 @@ public SNITCPHandle(string serverName, int port, long timerExpire, object callba
-        private static async Task<Socket> ConnectAsync(string serverName, int port)
+        private static Socket Connect(string serverName, int port, TimeSpan timeout)
         {
-            IPAddress[] addresses = await Dns.GetHostAddressesAsync(serverName).ConfigureAwait(false);
-            IPAddress targetAddrV4 = Array.Find(addresses, addr => (addr.AddressFamily == AddressFamily.InterNetwork));
-            IPAddress targetAddrV6 = Array.Find(addresses, addr => (addr.AddressFamily == AddressFamily.InterNetworkV6));
-            if (targetAddrV4 != null && targetAddrV6 != null)
+            IPAddress[] ipAddresses = Dns.GetHostAddresses(serverName);
+            IPAddress serverIPv4 = null;
+            IPAddress serverIPv6 = null;
+            foreach (IPAddress ipAdress in ipAddresses)
             {
-                return await ParallelConnectAsync(new IPAddress[] { targetAddrV4, targetAddrV6 }, port).ConfigureAwait(false);
+                if (ipAdress.AddressFamily == AddressFamily.InterNetwork)
+                {
+                    serverIPv4 = ipAdress;
+                }
+                else if (ipAdress.AddressFamily == AddressFamily.InterNetworkV6)
+                {
+                    serverIPv6 = ipAdress;
+                }
             }
-            else
-            {
-                IPAddress targetAddr = (targetAddrV4 != null) ? targetAddrV4 : targetAddrV6;
-                var socket = new Socket(targetAddr.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
+            ipAddresses = new IPAddress[] { serverIPv4, serverIPv6 };
+            Socket[] sockets = new Socket[2];
 
-                try
+            CancellationTokenSource cts = new CancellationTokenSource();
+            cts.CancelAfter(timeout);
+            void Cancel()
+            {
+                foreach (Socket socket in sockets)
                 {
-                    await socket.ConnectAsync(targetAddr, port).ConfigureAwait(false);
+                    if (socket != null && !socket.Connected)
+                    {
+                        socket.Dispose();
+                    }
                 }
-                catch
+            }
+            cts.Token.Register(Cancel);
+
+            Socket availableSocket = null;
+            for (int i = 0; i < sockets.Length; ++i)
+            {
+                try
                 {
-                    socket.Dispose();
-                    throw;
+                    if (ipAddresses[i] != null)
+                    {
+                        sockets[i] = new Socket(ipAddresses[i].AddressFamily, SocketType.Stream, ProtocolType.Tcp);
+                        sockets[i].Connect(ipAddresses[i], port);
+                        if (sockets[i].Connected)
+                        {
+                            availableSocket = sockets[i];
+                            break;
+                        }
+                    }
                 }
-                return socket;
+                catch { }
```

Comment:
Should the socket be disposed in case of the exception? Or are you relying on the cancel callback to take care of disposing the errored out sockets?

Reply:
Yes, it relies on the cancel callback to dispose failed sockets. Cancel callback method is kicked off anyway after certain time that configured by parameter, and it disposes any unsuccessful socket. However, I found edge case. If the cancel callback is executed the main thread creates a socket, and it is failed to connect, then the socket will not be disposed until destructor  is called. I will fix it.

Reply:
You can not rely on anything since the moment of catching System.Exception. Consider **OutOfMemoryException** happened in the constructor of Socket:
1) This socket will not be disposed until GC
2) Client does not know about exception

Even if we have considered all possible exceptions (including thrown by runtime, debuggers, profilers etc) we can not be sure it will not get changed. A developer trying to throw an exception somewhere inside does not know that code upper is not transparent.
https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1031?view=vs-2019
https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/exceptions/

Can we have an issue regarding swallowing the **System.Exception** ?

Reply:
Also this:

```
    if (sockets[i] != null) // sockets[i] can be null if cancel callback is executed during connect()
                        {
                            if (sockets[i].Connected)
```
is not synchornized with this:
```
if (sockets[i] != null && !sockets[i].Connected)
                        {
                            sockets[i].Dispose();
                            sockets[i] = null;
```
Which means that  "if (sockets[i].Connected)"  will throw NullReferenceException, which will be swallowed. This way, the socket stays in connected state while returning IPv6 socket to the client, correct?
I'm not sure how to correlate the situation with GC, but it looks like socket-leak.
@geleems @saurabh500

Reply:
@corivera @karelz May I please have your attention here? I believe this rule [ca1031](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1031?view=vs-2019) gets violated too many times in the core functionality of .NET. May be it does make sense to force the rule, preventing PRs from been merged?
Because it does not make sense to discuss it every time. I think it is easier to discuss the exceptions from that rule (like Task.SetException, AggregatedException, TargetInvocationException etc)